src/luascript.cpp

```diff
@@ -13777,66 +13812,927 @@ int LuaScriptInterface::luaPartySetSharedExperience(lua_State* L)
 // Spells
 int LuaScriptInterface::luaSpellCreate(lua_State* L)
 {
-	// Spell(words, name or id)
-	InstantSpell* spell = nullptr;
+	// Spell(words, name or id) to get an existing spell
+	// Spell(type) ex: Spell(SPELL_INSTANT) or Spell(SPELL_RUNE) to create a new spell
+	SpellType_t type = getNumber<SpellType_t>(L, 2);
+	if (isString(L, 2)) {
+		std::string tmp = asLowerCaseString(getString(L, 2));
+		if (tmp == "instant") {
+			type = SPELL_INSTANT;
+		} else if (tmp == "rune") {
+			type = SPELL_RUNE;
+		}
+	}
+
+	if (type) {
+		switch (type) {
+			case SPELL_INSTANT: {
+				InstantSpell* spell = new InstantSpell(getScriptEnv()->getScriptInterface());
+				if (spell) {
+					spell->fromLua = true;
+					pushUserdata<InstantSpell>(L, spell);
+					setMetatable(L, -1, "Spell");
+					spell->spellType = SPELL_INSTANT;
+				} else {
+					lua_pushnil(L);
+				}
+				break;
+			}
+			case SPELL_RUNE: {
+				RuneSpell* spell = new RuneSpell(getScriptEnv()->getScriptInterface());
+				if (spell) {
+					spell->fromLua = true;
+					pushUserdata<RuneSpell>(L, spell);
```

Comment:
Technically speaking, pushing a userdata of `RuneSpell` type and retrieving a different type (even if it's the base of `RuneSpell`) is UB, because userdata is stored/retrieved from a `void*`. You should make sure that the same type is pushed and retrieved.

Reply:
that's why I'm using `dynamic_cast` on a direct `Spell` pointer, the `Spell` pointer itself gives UB, when casting onto `InstantSpell` or `RuneSpell` I get the right values just fine, I couldn't find another way without splitting `RuneSpells` and `InstantSpells` entirely, if you have something better in mind just let me know.

Reply:
You should always push and get a `Spell` and cast afterwards:
```
pushUserdata<Spell>(L, rune_spell);
Spell* sp = getUserdata<Spell>(L, -1);
```

Reply:
fixed in [9e888a9](https://github.com/otland/forgottenserver/pull/2584/commits/9e888a96bb55d97ef102cb18a8c91e78178c8ecf)